lib/foreman/foreman_url_renderer.rb

```diff
@@ -23,6 +34,15 @@ def foreman_url(action = nil, params = {})
       host = self if @host.nil? && self.class < Host::Base
       template_proxy = host.try(:provision_interface).try(:subnet).try(:template_proxy)
 
+      # Parameters which must not be URL-encoded (e.g. iPXE synax ${xxx})
+      raw_string = ''
+      raw_params = params.delete(:raw)
+      if raw_params
+        # Make sure there is at least one parameter
+        params[:has_raw] = 'yes'
+        raw_params.each_pair { |k, v| raw_string += "&#{k}=#{v}" }
```

Comment:
How about this to avoid the `has_raw` parameter?
```suggestion
        raw_string += params.any? ? '&' : '?'
        raw_string += raw_params.map { |k, v| "#{k}=#{v}" }.join('&')
```

(untested)

Reply:
Definitely looks better. one more suggestion - maybe name the parameter `unencoded_params` so it's clear what it means?

Reply:
Good, I had to tune it a bit because there's token as well which is not part of params hash, but definitely looks better.

Reply:
perhaps we could do
```
params[:token] = host.token.value if (host.try(:build) && host.try(:token)
```
before-hand here instead of getting the token inside render_foreman_url?

Reply:
Sure, amended this change makes sense.